community/primitive-collections/src/main/java/org/neo4j/collection/pool/MarshlandPool.java

```diff
@@ -123,69 +115,58 @@ public void release( T obj )
+    @Override
+    public void close()
     {
         for ( LocalSlotReference slotReference : slotReferences )
         {
-            LocalSlot<T> slot = (LocalSlot) slotReference.get();
-            if ( slot != null )
+            T object = (T) slotReference.object;
+            if ( object != null )
             {
-                T obj = slot.object;
-                if ( obj != null )
+                slotReference.object = null;
+                LocalSlot<T> slot = (LocalSlot<T>) slotReference.get();
+                if ( slot != null )
                 {
                     slot.set( null );
-                    pool.release( obj );
                 }
+                slotReference.clear();
+                pool.release( object );
             }
         }
-
-        for ( LocalSlotReference<T> reference = (LocalSlotReference) objectsFromDeadThreads.poll();
-            reference != null;
-            reference = (LocalSlotReference) objectsFromDeadThreads.poll() )
-        {
-            T instance = reference.object;
-            if ( instance != null )
-            {
-                pool.release( instance );
-            }
-        }
-    }
-
-    public void close()
-    {
-        disposeAll();
+        slotReferences.clear();
     }
 
     /**
-     * This is used to trigger the GC to notify us whenever the thread local has been garbage collected.
+     * Container for the "puddle", the small local pool each thread keeps.
      */
-    private static class LocalSlotReference<T> extends WeakReference<LocalSlot>
+    private static class LocalSlot<T>
     {
+
         private T object;
+        private final LocalSlotReference slotWeakReference;
 
-        private LocalSlotReference( LocalSlot referent, ReferenceQueue<? super LocalSlot> q )
+        LocalSlot( ReferenceQueue<LocalSlot<T>> referenceQueue )
         {
-            super( referent, q );
+            slotWeakReference = new LocalSlotReference( this, referenceQueue );
+        }
+
+        public void set( T obj )
+        {
+            slotWeakReference.object = obj;
+            this.object = obj;
         }
     }
 
     /**
-     * Container for the "puddle", the small local pool each thread keeps.
+     * This is used to trigger the GC to notify us whenever the thread local has been garbage collected.
      */
-    private static class LocalSlot<T>
+    private static class LocalSlotReference<T> extends WeakReference<LocalSlot>
     {
         private T object;
-        private final LocalSlotReference phantomReference;
-
-        LocalSlot( ReferenceQueue<LocalSlot<T>> referenceQueue )
-        {
-            phantomReference = new LocalSlotReference( this, referenceQueue );
-        }
 
-        public void set( T obj )
+        private LocalSlotReference( LocalSlot referent, ReferenceQueue<? super LocalSlot> q )
         {
-            phantomReference.object = obj;
-            this.object = obj;
+            super( referent, q );
         }
```

Comment:
These two classes changed place and `phantomReference` was renamed to `slotWeakReference`?

Reply:
Yes, sorry for that reordering mess, it will be better to do those in a separate commit next time.
And yes phantom was renamed to a weak to avoid additional confusions.
And correct classes were exchanged to have a class declaration before it will be used as part of constructor parameter.